keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,202 @@
+- Automatically provide logic for developers (of core-types and CRDs) to
+  guarantee immutability of specific fields.
+
+### Non-Goals
+
+There are a few non-goals:
+- OpenAPI has a notion of "readOnly", but this is only meant to simplify schema,
+  and one is never supposed to send a readOnly value. This is different because
+  we want "writeOnce", not "readOnly".
+
+## Proposal
+
+This proposal attempts to create a concept of immutable and to visit some of
+what this can mean, but doesn't intend to propose all the possible
+semantics. The proposal should not close the door to further improvements.
+
+### Semantics
+
+We'll define immutable as "writeOnce", which means that the fields can only be
+set at creation time, and can never be updated after. Attempts to update an
+immutable field will result in an error (as opposed to being ignored), though we
+could potentially add that semantics later.
+
+It's also important to note that this concept is orthogonal to field ownership
+and list associativity. Ideally we would manage to keep the concepts entirely
+separate. e.g. An atomic list is a different concept from an immutable list.
+
+There are different semantics for a field to be immutable, especially if we
+consider non-leaf fields like lists and maps.
+
+#### Field selection
+
+Scalar fields have an obvious "selection" mechanism: they are either immutable
+or not. Things get more complicated for lists and maps, since we need to know if
+the flag applies to the list/map itself, to its members, or both.
+
+A few possible semantics are described here:
+
+- Non-recursive: For a list or a map, one can not remove or add new items, but
+  existing items can be modified.
+- Recursive: means that none of their field can change, and new fields are not
+  accepted.
+- Recursive with addition and/or deletion:
+  - New/deleted fields in maps and lists are tolerated.
+  - The added/deleted type would probably be "Recursive" after the first level
+    (addition/delete is not recursive).
+  - For associative lists, keys are already "immutable" in a way (similar to how
+    names are "immutables" for objects). Changing the keys would mean deleting
+    and creating a new item in the list, which would be permitted by these
+    rules.
+- Mutable: A field set as mutable could cancel the recursive immutability.
+
+#### On mutation of selected fields
+
+- Error: mutation of an immutable field returns an error and the request is
+  rejected. No mutation is performed at all.
+- Ignored: mutation of the field (and/or sub-fields) are completely ignored, and
+  not persisted. They do not trigger an error.
+
+#### Equality
+
+Equality of two fields is very important:
+- Scalar fields are compared directly: fields and values must exactly match from old
+  object to new object.
+- Lists fields are compared using the same logic currently used for native
+  kubernetes types: Empty and null lists are equal. A missing list is NOT equal
+  to a null or empty list.
```

Comment:
I would really like to have a way in CRD OpenAPI to specify a normalization behaviour for lists and maps:

```
x-kubernetes-normalize-null: empty
x-kubernetes-normalize-null: undefined
x-kubernetes-normalize-empty: null
x-kubernetes-normalize-empty: undefined
```

And then we should be strict here in the mutation test.

Also note that in Proto we have neither null nor empty by default. I still think we need a more broad discussion of null, undefined, empty values.

Reply:
I don't disagree, I don't know if this should be blocking?

Reply:
At least we have to leave it open here and have a plan before marking it implementable. But I like to move forward with implementation before being very clear about this topic and where it is heading.

Reply:
Or in other words: whatever we decide here will drive the API design of the whole CRD ecosystem. We should take the time it needs to understand the topic.

Reply:
I think we should make a decision and enforce it rather than offer too many complicated options. We currently consider empty and null to be equal for existing type, and I think that we should just tell people that's how things are going to work.

Reply:
We decided against that for defaulting. Also in validation null is distinguishable from undefined. It would be very odd to have immutable checks to differ.

Reply:
I'd like to reboot this discussion.

Please let me know if I'm missing something important, but when reviewing defaulting, validation and equality, it looks like we have:

**`Semantic.DeepEqual`:**

| ==               | missing field | null | empty collection |
|------------------|---------------|------|------------------|
| missing field    | yes           | no   | no               |
| null             | no            | yes  | yes              |
| empty collection | no            | yes  | yes              |


**Defaulting:**

Applied only to missing fields. It does not default null or empty collections or have any opinion about null vs. empty.

**Validation:**

Has "nullable" rule.

  - non-nullable fields have strict equality.
  - nullable fields have "null == empty collecion" equality. 

So...

If immutability had "null == empty collecion" equality when checking for mutations, then we are allowing for field values that readers can distinguish between (any client can choose, if they want to, to handle null differently than an empty collection) to be considered equal throughout the system for the long term (forever?).

If immutability had strict equality for mutations, than we really need to update `Semantic.DeepEqual` and whatever other equality checks we have to match. Is this a API breaking change?

Is the logical next step to review `Semantic.DeepEqual` to see if that is something that is eligible for being changed?

Reply:
Have to remember where we left off with the disucssion. Compare https://github.com/apelisse/enhancements/pull/1 where we went into detail more already.

Reply:
As markdown tables are so tedious, I started https://docs.google.com/spreadsheets/d/1KpX_ZAgw8LKQMsgOV-UMhD2THzjljAGN31_8lCRquLA/edit?usp=sharing.

Reply:
My main problem with any non-strict behaviour for CRDs is that it feels very odd to have different objects in etcd, but consider them equal. 

For native resource with proto encoding this is not the because proto unifies undefined, null and empty all to one value und unmarshal those to null for comparison.

I wonder whether we should talk about normalization for CRDs before talking about equality:

- we have one normalization already: undefined => default value
- we don't have normalization for empty
- we don't have normalization for null

With normalization happening as part of our deserialization stack, we can separate both concerns and use strict equality in immutability checks.

Today we use semantic equality mostly because JSON and Proto have different normalization behaviours (partly driven by omitempty as well).

| normalization JSON omitempty | |
| - | - |
| undefined | undefined |
| null | undefined |
| empty | undefined |

| normalization JSON | |
| - | - |
| undefined | null |
| null | null |
| empty | empty |

| normalization Proto omitempty/non-omitempty | |
| - | - |
| undefined | undefined |
| ~~null~~ | ~~undefined~~ |
| ~~empty~~ | ~~undefined~~ |

| normalization CRD omitempty/non-omitempty | |
| - | - |
| undefined | undefined |
| null | null |
| empty | empty |

| normalization CRD default=null | |
| - | - |
| undefined | null |
| null | null |
| empty | empty |

| normalization CRD default=empty | |
| - | - |
| undefined | empty |
| null | null |
| empty | empty |

Reply:
If I'm understanding all the normalization rules right, then for native types we have something like this?

| normalization native, client speaking JSON | non-omitempty | omitempty |
|--------------------------------|-----------|-----------|
| undefined                      | empty | undefined |
| null                           | empty | undefined |
| empty                          | empty | undefined |

| normalization native omitempty/non-omitempty, client speaking Proto | |
|--------------------------------|-----------|
| undefined                      | undefined |
| null                           | undefined |
| empty                          | undefined |

And we don't have omitempty empty for CRDs?

Would it make sense to normalize the same way for CRDs as we do for native types (using proto)? I.e. normalize null and empty to undefined. We'd need to normalize all data received from clients and all data read from etcd (since we have non-normalized data at rest). If we did that, we should be able to do strict equality in the api-server code.

I'm guesses that to be consistent we should also normalize native types stored as json like we do for CRDs?

Reply:
not sure about the first table, what are the values?

> And we don't have the concept of omitempty empty for CRDs?

There is no Golang type that could have omitempty in the request handler logic. JSON unmarshalling for CRD is faithfully, i.e. does not swallow or normalize anything.

I am not convinced we want to start putting up fuzzy glasses for CRDs by adding normalization by default. Moreover, we cannot even do that without breaking compatibility, at least not in general (we could do it for equality only).

I don't like the idea of having the normal REST logic with strict empty-null-undefined logic, but immutability to act differently. That's pretty counter-intuitive.

I tend to think about normalization as a property of the fields, i.e. the schema properties. For example we can express nullable fields (which preserve null in all cases, and empty and undefined) using special Golang types (like extra pointers) and at worst custom marshallers. What we don't do today is publishing of this as part of OpenAPI, and in general we never formalized that. But my point is that there are native types which do not normalize prior to persisting values and before comparing, it is just not what we get by default with naive or lazy Golang type definitions and our serializer stack.

In other words, we might want to formalize the normalization of any slice and map in the schemas, for native types and for CRDs. CRD schemas could get a `x-kubernetes-normalize-empty: undefined` vendor extensions (not necessarily for 1.17, could be added later). Equality should be strict, relative to the defined normalization (which is not the indenty normalization for native types then).

Also for later introduction of protobuf support for a CRDs, we can require a certain normalization that is compatible with proto. I.e. the CRD validation of a CRD instance with proto enabled, would validate the proto numerals are all specified, and it would validate that normalization is compatible.

Reply:
```
properties:
  foo:
    x-kubernetes-proto: 6
    x-kubernetes-normalize-empty: empty
    x-kubernetes-normalize-null: empty # if nullable
    x-kubernetes-normalize-undefined: undefined
    nullable: true
```

`validation error: properties.foo.x-kubernetes-normalize-undefined must be empty if x-kubernetes-proto is set.`

Reply:
Thanks for explaining. I don't think I appreciated the extend our existing golang type definitions play into the situation. That's good to know.

From a usability perspective, it wish we could provide a sane default normalization rule that users could get by default and not have to specify any normalization settings explicitly, but I agree we can't just start doing that for CRDs since it breaks backward compatibility.

I really like the idea of specifying the normalization we expect then validating our encodings and go types are all aligned with that.